### PR TITLE
[SPARK-19641][SQL] JSON schema inference in DROPMALFORMED mode produces incorrect schema for non-array/object JSONs

### PR DESCRIPTION
**Pull request by HyukjinKwon**
## What changes were proposed in this pull request?

Currently, when we infer the types for vaild JSON strings but object or array, we are producing empty schemas regardless of parse modes as below:

```scala
scala> spark.read.option("mode", "DROPMALFORMED").json(Seq("""{"a": 1}""", """"a"""").toDS).printSchema()
root
```

```scala
scala> spark.read.option("mode", "FAILFAST").json(Seq("""{"a": 1}""", """"a"""").toDS).printSchema()
root
```

This PR proposes to handle parse modes in type inference.

After this PR,

```scala

scala> spark.read.option("mode", "DROPMALFORMED").json(Seq("""{"a": 1}""", """"a"""").toDS).printSchema()
root
 |-- a: long (nullable = true)
```

```
scala> spark.read.option("mode", "FAILFAST").json(Seq("""{"a": 1}""", """"a"""").toDS).printSchema()
java.lang.RuntimeException: Failed to infer a common schema. Struct types are expected but string was found.
```

This PR is based on https://github.com/NathanHowell/spark/commit/e233fd03346a73b3b447fa4c24f3b12c8b2e53ae and I and @NathanHowell talked about this in https://issues.apache.org/jira/browse/SPARK-19641

## How was this patch tested?

Unit tests in `JsonSuite` for both `DROPMALFORMED` and `FAILFAST` modes.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @NathanHowell and @cloud-fan.

**Comment by HyukjinKwon** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonInferSchema.scala [hunk 1]
```diff
@@ -217,26 +221,43 @@ private[sql] object JsonInferSchema {
     }
   }
 
+  private def withParseMode(
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonInferSchema.scala]: shall we embed this method in `withCorruptField`?

Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonInferSchema.scala]: Sure.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonInferSchema.scala [hunk 2]
```diff
@@ -202,41 +206,54 @@ private[sql] object JsonInferSchema {
...
+        // Otherwise, just return this struct.
+        struct
+      }
+
+    case DropMalformedMode =>
+      // If corrupt record handling is disabled we retain the valid schema and discard the other.
       struct
-    }
+
+    case FailFastMode =>
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonInferSchema.scala]: When will we hit this branch? Seems never?

Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonInferSchema.scala]: It looks possible to run this line. I added a test in 1936L. In more details, if the json is a valid one but not a object of an areay of object, it will infer not `StructType` per record. If one of the types is not a struct type and failfast mode is enabled, we will hit this line.

I am outside now. I will double check when I get to my computer.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala [hunk 3]
```diff
@@ -1903,9 +1932,8 @@ class JsonSuite extends QueryTest with SharedSQLContext with TestJsonData {
           .option("wholeFile", true)
           .option("mode", "FAILFAST")
           .json(path)
-          .collect()
       }
-      assert(exceptionOne.getMessage.contains("Failed to parse a value"))
+      assert(exceptionOne.getMessage.contains("Failed to infer a common schema"))
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala]: Up to my knowledge, this test case throws an exception when actually parsing the data before. Now, I intended to check the exception in schema inference as it looks parsing data case is covered below.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala [hunk 4]
```diff
@@ -1041,7 +1041,6 @@ class JsonSuite extends QueryTest with SharedSQLContext with TestJsonData {
       spark.read
         .option("mode", "FAILFAST")
         .json(corruptRecords)
-        .collect()
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala]: I intended to remove this line to make sure it happens in schrma inference. I think this is not related with the change here.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by HyukjinKwon** [comment]
Thank you @cloud-fan.
